File: lib/java/openlinktoken/src/main/java/org/openlinktoken/tokentransformer/DecryptTokenTransformer.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokentransformer;

import java.nio.charset.StandardCharsets;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.Base64;
import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Transforms the token using AES-256 symmetric decryption.
 *
 * @see <a href=https://datatracker.ietf.org/doc/html/rfc3826>AES</a>
 */
public class DecryptTokenTransformer implements TokenTransformer {
    private static final long serialVersionUID = 1L;

    private static final Logger logger = LoggerFactory.getLogger(DecryptTokenTransformer.class);

    private final SecretKeySpec secretKey;

    /**
     * Initializes the underlying cipher (AES) with the decryption secret.
     *
     * @param encryptionKey the encryption key. The UTF-8 encoded key material must be exactly 32 bytes long.
     *
     * @throws InvalidKeyException                invalid encryption key
     * @throws InvalidAlgorithmParameterException invalid encryption algorithm
     *                                             parameters
     */
    public DecryptTokenTransformer(String encryptionKey)
            throws InvalidKeyException, InvalidAlgorithmParameterException {
        this(toValidatedKeyBytes(encryptionKey));
    }

    /**
     * Initializes the underlying cipher (AES) with raw decryption key material.
     *
     * @param encryptionKey the raw encryption key bytes. The key must be exactly 32 bytes long.
     * @throws InvalidKeyException                invalid encryption key
     * @throws InvalidAlgorithmParameterException invalid encryption algorithm
     *                                             parameters
     */
    public DecryptTokenTransformer(byte[] encryptionKey)
            throws InvalidKeyException, InvalidAlgorithmParameterException {
        this.secretKey = new SecretKeySpec(toValidatedKeyBytes(encryptionKey), EncryptionConstants.AES);
    }

    /**
     * Converts a string key to UTF-8 bytes and validates its length.
     *
     * @param encryptionKey the string encryption key
     * @return a defensive copy of the validated key bytes
     * @throws InvalidKeyException if the key is not exactly 32 bytes long
     */
    private static byte[] toValidatedKeyBytes(String encryptionKey) throws InvalidKeyException {
        return toValidatedKeyBytes(encryptionKey == null ? null : encryptionKey.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Validates and copies raw encryption key material.
     *
     * @param encryptionKey the raw encryption key bytes
     * @return a defensive copy of the validated key bytes
     * @throws InvalidKeyException if the key is not exactly 32 bytes long
     */
    private static byte[] toValidatedKeyBytes(byte[] encryptionKey) throws InvalidKeyException {
        if (encryptionKey == null || encryptionKey.length != EncryptionConstants.KEY_BYTE_LENGTH) {
            logger.error("Invalid Argument. Key must be {} bytes long", EncryptionConstants.KEY_BYTE_LENGTH);
            throw new InvalidKeyException(String.format("Key must be %s bytes long", EncryptionConstants.KEY_BYTE_LENGTH));
        }
        return Arrays.copyOf(encryptionKey, encryptionKey.length);
    }

    /**
     * Decryption token transformer.
     * <p>
     * Decrypts the token using AES-256 symmetric decryption algorithm.
     *
     * @return the decrypted token string.
     * @param token the encrypted token in base64 format.
     * @throws IllegalArgumentException        if the token does not contain the
     *                                         required IV and authentication tag
     * @throws IllegalStateException        if the underlying cipher is in an
     *                                      invalid state
     * @throws IllegalBlockSizeException    if the encrypted input cannot be
     *                                      processed as a complete block
     * @throws BadPaddingException          if the encrypted input has invalid
     *                                      padding or authentication data
     * @throws InvalidAlgorithmParameterException if the encryption parameters are
     *                                      invalid
     * @throws InvalidKeyException          if the encryption key is invalid
     * @throws NoSuchAlgorithmException     if the encryption algorithm is not
     *                                      available
     * @throws NoSuchPaddingException       if the encryption padding is not
     *                                      available
     */
    @Override
    public String transform(String token)
            throws IllegalStateException, IllegalBlockSizeException, BadPaddingException, InvalidKeyException,
            InvalidAlgorithmParameterException, NoSuchAlgorithmException, NoSuchPaddingException {
        // Decode the base64-encoded token
        byte[] messageBytes = Base64.getDecoder().decode(token);

        int minimumMessageLength = EncryptionConstants.IV_SIZE + EncryptionConstants.TAG_LENGTH_BITS / Byte.SIZE;
        if (messageBytes.length < minimumMessageLength) {
            throw new IllegalArgumentException(
                    "Encrypted token is missing its initialization vector or authentication tag");
        }

        // Extract IV and ciphertext
        byte[] ivBytes = Arrays.copyOfRange(messageBytes, 0, EncryptionConstants.IV_SIZE);
        byte[] cipherBytes = Arrays.copyOfRange(messageBytes, EncryptionConstants.IV_SIZE, messageBytes.length);

        GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(EncryptionConstants.TAG_LENGTH_BITS, ivBytes);

        // Initialize AES cipher in GCM mode with no padding for decryption
        Cipher cipher = Cipher.getInstance(EncryptionConstants.ENCRYPTION_ALGORITHM);
        cipher.init(Cipher.DECRYPT_MODE, this.secretKey, gcmParameterSpec);

        byte[] decryptedBytes = cipher.doFinal(cipherBytes);

        return new String(decryptedBytes, StandardCharsets.UTF_8);
    }
}

File: lib/python/openlinktoken/src/main/openlinktoken/tokentransformer/decrypt_token_transformer.py
# SPDX-License-Identifier: MIT

import base64
import logging
from typing import Union

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from openlinktoken.tokentransformer.encryption_constants import EncryptionConstants
from openlinktoken.tokentransformer.token_transformer import TokenTransformer

logger = logging.getLogger(__name__)


class DecryptTokenTransformer(TokenTransformer):
    """AES-256 GCM token decryption transformer.

    Parity with Java `DecryptTokenTransformer`:
    - Java prepends IV to (ciphertext || tag) and GCM consumes combined remainder.
    - Python splits IV, ciphertext, and tag explicitly before constructing GCM mode.
    - Both expect 12-byte IV and 16-byte tag; key length enforced at init.
    """

    def __init__(self, encryption_key: Union[str, bytes]):
        """
        Initializes the underlying cipher (AES) with the decryption secret.

        Accepts either a ``str`` (UTF-8 encoded; must encode to exactly 32 bytes) or
        raw ``bytes`` (must be exactly 32 bytes).

        Args:
            encryption_key: The encryption key. The UTF-8 encoded key material must be exactly 32 bytes long.

        Raises:
            ValueError: If the encryption key material is not exactly 32 bytes long.
        """
        if isinstance(encryption_key, bytes):
            if len(encryption_key) != EncryptionConstants.KEY_BYTE_LENGTH:
                logger.error(f"Invalid Argument. Key must be {EncryptionConstants.KEY_BYTE_LENGTH} bytes long")
                raise ValueError(f"Key must be {EncryptionConstants.KEY_BYTE_LENGTH} bytes long")
            self.encryption_key = encryption_key
        else:
            encryption_key_bytes = encryption_key.encode("utf-8")
            if len(encryption_key_bytes) != EncryptionConstants.KEY_BYTE_LENGTH:
                logger.error(f"Invalid Argument. Key must be {EncryptionConstants.KEY_BYTE_LENGTH} bytes long")
                raise ValueError(f"Key must be {EncryptionConstants.KEY_BYTE_LENGTH} bytes long")
            self.encryption_key = encryption_key_bytes

    def transform(self, token: str) -> str:
        """Decrypt a base64 encoded token produced by the encrypt transformer.

        Args:
            token: Base64 string containing IV || ciphertext || tag.

        Returns:
            Decrypted UTF-8 token string.

        Raises:
            Exception: Propagates cryptographic failures (logged here first).
        """
        try:
            # Decode the base64-encoded token
            message_bytes = base64.b64decode(token)

            minimum_message_length = EncryptionConstants.IV_SIZE + EncryptionConstants.TAG_LENGTH_BYTES
            if len(message_bytes) < minimum_message_length:
                raise ValueError("Encrypted token is missing its initialization vector or authentication tag")

            # Extract IV, encrypted data, and tag
            iv_bytes = message_bytes[: EncryptionConstants.IV_SIZE]
            ciphertext_and_tag = message_bytes[EncryptionConstants.IV_SIZE :]
            ciphertext = ciphertext_and_tag[: -EncryptionConstants.TAG_LENGTH_BYTES]
            tag = ciphertext_and_tag[-EncryptionConstants.TAG_LENGTH_BYTES :]

            # Create cipher for decryption
            cipher = Cipher(algorithms.AES(self.encryption_key), modes.GCM(iv_bytes, tag), backend=default_backend())

            # Decrypt the token
            decryptor = cipher.decryptor()
            decrypted_bytes = decryptor.update(ciphertext) + decryptor.finalize()

            return decrypted_bytes.decode("utf-8")

        except Exception as e:
            logger.error("Error during token decryption", exc_info=e)
            raise
